pkg/p4rtclient/conn_manager.go

```diff
@@ -0,0 +1,308 @@
+// SPDX-FileCopyrightText: 2022-present Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package p4rtclient
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"github.com/google/uuid"
+	topoapi "github.com/onosproject/onos-api/go/onos/topo"
+	"github.com/onosproject/onos-lib-go/pkg/certs"
+	"github.com/onosproject/onos-lib-go/pkg/errors"
+	"github.com/onosproject/onos-lib-go/pkg/logging"
+	p4v1 "github.com/p4lang/p4runtime/go/p4/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"math"
+	"sync"
+	"time"
+)
+
+var log = logging.GetLogger()
+
+const (
+	defaultTimeout = 60 * time.Second
+)
+
+// ConnManager p4rt connection manager
+type ConnManager interface {
+	Get(ctx context.Context, connID ConnID) (Conn, bool)
+	GetByTarget(ctx context.Context, targetID topoapi.ID) (Client, error)
+	Connect(ctx context.Context, destination *Destination) error
+	Disconnect(ctx context.Context, targetID topoapi.ID) error
+	Watch(ctx context.Context, ch chan<- Conn) error
+}
+
+// NewConnManager creates a new p4rt connection manager
+func NewConnManager() ConnManager {
+	mgr := &connManager{
+		targets:  make(map[topoapi.ID]*client),
+		conns:    make(map[ConnID]Conn),
+		watchers: make(map[uuid.UUID]chan<- Conn),
+		eventCh:  make(chan Conn),
+	}
+	go mgr.processEvents()
+	return mgr
+}
+
+type connManager struct {
+	targets    map[topoapi.ID]*client
+	conns      map[ConnID]Conn
+	connsMu    sync.RWMutex
+	watchers   map[uuid.UUID]chan<- Conn
+	watchersMu sync.RWMutex
+	eventCh    chan Conn
+}
+
+// Get returns a connection based on a given ID
+func (m *connManager) Get(ctx context.Context, connID ConnID) (Conn, bool) {
+	m.connsMu.RLock()
+	defer m.connsMu.RUnlock()
+	conn, ok := m.conns[connID]
+	return conn, ok
+}
+
+// GetByTarget returns a P4RT client based on a given target ID
+func (m *connManager) GetByTarget(ctx context.Context, targetID topoapi.ID) (Client, error) {
+	m.connsMu.RLock()
+	defer m.connsMu.RUnlock()
+	if p4rtClient, ok := m.targets[targetID]; ok {
+		return p4rtClient, nil
+	}
+	return nil, errors.NewNotFound("p4rt client for target %s not found", targetID)
+}
+
+// Connect makes a gRPC connection to the target
+func (m *connManager) Connect(ctx context.Context, destination *Destination) error {
```

Comment:
User should provide a target ID otherwise endpoint address (ip, port) will be used for target ID that we can keep track of connections to one or multiple targets. Is this a good assumption? @tomikazi